### PR TITLE
Speed up local audio

### PR DESCRIPTION
**Pull request by zzmp**

### CODE REVIEW

**Comment by zzmp** [comment]
Waiting on #9935.

**Comment by zzmp** [comment]
#### QA Plan

AUDIO-LATENCY-03 Test
AUDIO-LATENCY-04 Test

**Comment by zzmp** [comment]
I'm unable to connect to Atlassian right now. AUDIO-LATENCY-04 Test is as follow:

**Objective: Local injectors should be queued as quickly as possible. This test is meant to benchmark master against a PR, not to establish that a PR is functioning at all.**

**Precondition: Enter any domain with limited audio.**

```js
// Load sounds
var A2 = SoundCache.getSound("http://mpassets.highfidelity.com/3fd92e5e-93cf-4bc1-b2f1-c6bae5629814-v1/A2.wav");
// Wait a moment for the sounds to load, then...

// Time how long it takes 5000 sounds to play.
// Ideally, they would all start immediately, and it would only take the duration of a single sound.
// Actually, between JavaScript and memory allocations, this will take upwards of 15s.
for (var i = 0; i < 5000; i++) Audio.playSound(A2, { position: { x: 0, y: 0, z: 0 }, volume: 1 });
// For this test to pass, the PR should take less time than the master branch.
```

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8146ab5528806dcd38b88875a9384130c930f2a0 available at: https://gist.github.com/4e38110f66157a427a59ba3d05e0de19<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9936-8146ab5.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9936-8146ab5.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8c563881c6d216458313d67f550dab8683206add available at: https://gist.github.com/4e38110f66157a427a59ba3d05e0de19<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9936-8c56388.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9936-8c56388.dmg

File: libraries/audio-client/src/AudioClient.cpp [hunk 1]
```diff
@@ -1522,14 +1516,23 @@ bool AudioClient::switchOutputToAudioDevice(const QAudioDeviceInfo& outputDevice
...
 
                     // size local output mix buffer based on resampled network frame size
                     _networkPeriod = _localToOutputResampler->getMaxOutput(AudioConstants::NETWORK_FRAME_SAMPLES_STEREO);
                     _localOutputMixBuffer = new float[_networkPeriod];
+
+                    // local period should be at least twice the output period,
+                    // in case two device reads happen before more data can be read (worst case)
                     int localPeriod = _outputPeriod * 2;
+                    // it should also be full network periods so that it actually fills
+                    localPeriod += _networkPeriod - localPeriod % _networkPeriod;
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: If localPeriod is already a multiple of _networkPeriod, this adds an extra _networkPeriod to the buffer.

A more idiomatic way to compute this (and avoids the problem) is
```
int localPeriod = _outputPeriod * 2;
// round up to an exact multiple of _networkPeriod
localPeriod = ((localPeriod + _networkPeriod-1) / _networkPeriod) * _networkPeriod;
```

**Comment by kencooke** [comment]
### Additional QA:

We should smoke-test this on at least one other platform (Linux or Mac) since the different audio backend will likely result in different callback size/behavior. If a sinewave injector plays back cleanly, PASS.

**Comment by OmegaHeron** [comment]
Testing Windows 10 and Ubuntu 16.04 now.

**Comment by OmegaHeron** [comment]
@zzmp - It's unclear to me if there's something missing in test definition or if I need to manually time something.  Ping via Slack or here, I'll hold testing until.  Thanks.

**Comment by OmegaHeron** [comment]
I think I see what you did now - use timestamps in log?

`[03/21 13:49:33] [DEBUG] [hifi.networking] Finished loading: http://mpassets.highfidelity.com/3fd92e5e-93cf-4bc1-b2f1-c6bae5629814-v1/A2.wav`

`[03/21 13:49:50] [DEBUG] [hifi.audioclient] adding new injector`
...
`[03/21 13:50:10] [DEBUG] [hifi.audioclient] removing injector`

For 20 seconds from time sound loop play begins to last indication of removing injector?

**Comment by OmegaHeron** [comment]
Passing this back to QA Pool - out of time for today.

**Comment by zzmp** [comment]
Nope, I used a stopwatch manually.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0d5a017fe133d808d3518113f2ed833d0c83639c available at: https://gist.github.com/4e38110f66157a427a59ba3d05e0de19<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9936-0d5a017.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9936-0d5a017.dmg

**Comment by OmegaHeron** [comment]
Tested Windows 10
AUDIO-LATENCY-03 Test
Master-6208 *Pass*
PR *Pass*

AUDIO-LATENCY-04 Test
Master-6208 5.5 Seconds.
PR 5.5 Seconds.

How I tested.  I opened console, copy/pasted line to load sound into A2 var.  Waited until it was loaded as shown by downloads stat in expanded stats display.  Then pasted second line into console and started stopwatch precisely as I press enter to fire sound test.  Repeated 5x each in master vs PR in a domain where no sounds playing.  So - really I can't discern any difference between PR/Master and certainly nothing approaching 20 seconds.  CPU = i5 6500.  @zzmp, Please advise/clarify/confirm AUDIO-LATENCY-04 Test is correct and that I'm engaging it proper.

**Comment by zzmp** [comment]
@OmegaHeron I'll retest on my end. I also think that your idea of using timestamps may be better / more accurate. It's strange that we should get such different results - maybe its because of the commit I made after ken's comments.

**Comment by PhilipRosedale** [comment]
@zzmp what is the status on this PR?

**Comment by zzmp** [comment]
Super stale, sorry.
This PR changed two things - the condition under which local audio was buffered, and the size of the buffer.
The condition has since been clarified, so I am closing this PR because it is no longer relevant.
The size should still be changed, but I will open up a new PR to do that.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5030d60682f482180a026cc8c9832ca49e0e5cbb available at: https://gist.github.com/4e38110f66157a427a59ba3d05e0de19<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9936-5030d60.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9936-5030d60.dmg

**Pull request closed without merging**